cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/artifact/RemotePluginFinder.java

```diff
@@ -209,10 +217,74 @@ public Location getArtifactLocation(ArtifactId artifactId)
 
     String path = response.getResponseBodyAsString();
     Location location = Locations.getLocationFromAbsolutePath(locationFactory, path);
+
+    // If the artifact doesn't exist locally then fetch it from app fabric and save it locally
     if (!location.exists()) {
-      throw new IOException(String.format("Artifact Location does not exist %s for artifact %s version %s",
-                                          path, artifactId.getArtifact(), artifactId.getVersion()));
+      LOG.debug(String.format("Artifact '%s' is not present locally at %s, attempting to fetch it from app-fabric.",
+                              artifactId.getArtifact(), path));
+
+      Retries.runWithRetries(() -> getAndStoreArtifact(artifactId, location), retryStrategy);
     }
     return location;
   }
+
+  private void getAndStoreArtifact(ArtifactId artifactId, Location location) throws IOException {
```

Comment:
Is the location supposed for caching? It seems quite raw to implement like this. Is this logic going to be replaced with the file fetcher?

Reply:
Yes, this is temporary until the file fetcher sidecar work is completed